src/service/visibility-impl.js

```diff
@@ -0,0 +1,134 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {getService} from '../service';
+import {resourcesFor} from '../resources';
+import {viewportFor} from '../viewport';
+
+/**
+ * This type signifies a callback that gets called when visibility conditions
+ * are met.
+ * @typedef {function()}
+ */
+let VisibilityListenerCallbackDef;
+
+/**
+ * @typedef {Object.<string, JSONObject|VisibilityListenerCallbackDef>}
```

Comment:
The `|` is the confusing part then. Can you create a `VisibilityListenerConfigDef`?